### DISCUSSION TITLE
Should with_prefix be used for Unit type enums?

### DISCUSSION THREAD

**Discussion by elpiel** [jonasbb/serde_with#204] [Q&A] [answered]
I have a special use case where I use an Enum as a way to define fields of a map which is dynamically build. Since serde can represent the value `MyEnum::Field` as `"Field"` I was also looking for a way to use a prefix for that field.

Right now this examples doesn't compile

```rust
serde_with::with_prefix!(my_enum_prefix "myEnum.");

#[derive(Debug, Serialize, Deserialize)]
enum MyEnum {
    #[serde(with = "my_enum_prefix")]
    Field,
    #[serde(with = "my_enum_prefix")]
    Field2
}
```

Compile error:

```
error[E0061]: this function takes 2 arguments but 1 argument was supplied
   --> src/lib.rs:157:40
    |
9   | serde_with::with_prefix!(adview_prefix "adView.");
    | -------------------------------------------------- defined here
...
157 |     #[derive(Hash, Copy, Clone, Debug, Serialize, Deserialize, Eq, PartialEq)]
    |                                        ^^^^^^^^^
    |                                        |
    |                                        supplied 1 argument
    |                                        expected 2 arguments
    |
    = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)
```

I also tried to add it to the enum of fields:

```rust
#[derive(Debug, Serialize, Deserialize)]
#[serde(untagged)]
pub enum Field {
        AdView(#[serde(with = "...")] AdView),
....
}
```

However this results to the prefix not being accounted for.

**Comment by jonasbb** [comment, marked as answer]
The `with_prefix!` macro is meant to be used to add prefixes to a struct/enum, if you cannot fully control the `Serialize` implementation or if you need different prefixes. Something like this:

```rust
struct SomeStruct {
	// ...
	#[serde(with = "my_enum_prefix")]
	my_enum: MyEnum,
	// ..
}
```

I don't think `with_prefix!` can really be used to rename the fields of a struct or unit variants without such a wrapping struct/enum. The field name / unit variant already gets emitted before the with part runs, so it cannot change those values.

The solution right now is to use `#[serde(rename = "myEnum....")]` on every enum variant.

Alternatively, it should be possible to provide a proc-macro which performs this step. It could be used like:

```rust
#[serde_with::prefix_all("my_enum")]
enum MyEnum {
	// ...
}
```

Ideally it would work on enum variants and struct fields.

**Comment by elpiel** [comment]
Thanks for the feedback and indeed a proc-macro could work.
I am ok with what I can use right now, but I thought I would ask anyway
